src/Security/AzureAuthenticator.php

```diff
@@ -0,0 +1,140 @@
+
+use App\DTO\UserDto;
+use App\Entity\User;
+use App\Repository\UserRepository;
+use App\Service\IpResolver;
+use App\Service\SettingsManager;
+use App\Service\UserManager;
+use App\Utils\Slugger;
+use Doctrine\ORM\EntityManagerInterface;
+use KnpU\OAuth2ClientBundle\Client\ClientRegistry;
+use KnpU\OAuth2ClientBundle\Security\Authenticator\OAuth2Authenticator;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\Routing\RouterInterface;
+use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
+use Symfony\Component\Security\Core\Exception\AuthenticationException;
+use Symfony\Component\Security\Core\Exception\CustomUserMessageAuthenticationException;
+use Symfony\Component\Security\Http\Authenticator\Passport\Badge\RememberMeBadge;
+use Symfony\Component\Security\Http\Authenticator\Passport\Badge\UserBadge;
+use Symfony\Component\Security\Http\Authenticator\Passport\Passport;
+use Symfony\Component\Security\Http\Authenticator\Passport\SelfValidatingPassport;
+use TheNetworg\OAuth2\Client\Provider\AzureResourceOwner;
+
+class AzureAuthenticator extends OAuth2Authenticator
+{
+    public function __construct(
+        private readonly ClientRegistry $clientRegistry,
+        private readonly RouterInterface $router,
+        private readonly EntityManagerInterface $entityManager,
+        private readonly UserManager $userManager,
+        private readonly IpResolver $ipResolver,
+        private readonly Slugger $slugger,
+        private readonly UserRepository $userRepository,
+        private readonly SettingsManager $settingsManager,
+    ) {
+    }
+
+    public function supports(Request $request): ?bool
+    {
+        return 'oauth_azure_verify' === $request->attributes->get('_route');
+    }
+
+    public function authenticate(Request $request): Passport
+    {
+        $client = $this->clientRegistry->getClient('azure');
+        $slugger = $this->slugger;
+
+        $provider = $client->getOAuth2Provider();
+
+        $accessToken = $provider->getAccessToken('authorization_code', [
+            'code' => $request->query->get('code'),
+        ]);
+
+        $rememberBadge = new RememberMeBadge();
+        $rememberBadge = $rememberBadge->enable();
+
+        return new SelfValidatingPassport(
+            new UserBadge($accessToken->getToken(), function () use ($accessToken, $client, $slugger) {
+                /** @var AzureResourceOwner $azureUser */
+                $azureUser = $client->fetchUserFromToken($accessToken);
+
+                $existingUser = $this->entityManager->getRepository(User::class)->findOneBy(
+                    ['oauthAzureId' => $azureUser->getUpn()]
+                );
+
+                if ($existingUser) {
+                    return $existingUser;
+                }
+
+                $user = $this->userRepository->findOneBy(['email' => $azureUser->getUpn()]);
```

Comment:
I'm not quite sure on this line, I guess I'm not familiar with what `Upn` is. Reading up the [docs I could find on it](https://learn.microsoft.com/en-us/entra/identity/hybrid/connect/plan-connect-userprincipalname)

This line is looking for an already created native account with an email value that is the same as an azure `Upn`. Whether that is ever possible, I'm not sure, but it's also possible it doesn't harm anything. It seems like maybe Upn _can_  be an email, so maybe this is alright? I just wanted to check, it doesn't seem like the [AzureResourceOwner](https://github.com/TheNetworg/oauth2-azure/blob/master/src/Provider/AzureResourceOwner.php) has a field for email or anything so I guess this is the most right it could be

Reply:
My comment might've seemed a bit rambling so my actual question is just, does this part seem ok to you, @WebVoid ?

Let's say your Upn is `bob`, the only way you'd ever get this to run is most likely creating an account via CLI with the email `bob` (since you wouldn't be able to via registration on the site as it's not a valid email), and then it would enter this if, find that user, and assign their `oauthAzureId` to `bob`. Probably fine?

Reply:
As for the tests I have done, the upn was the email of the person authenticating. I remember trying to look for a better field as an identifier but couldn't find any. I could var_dump the object returned by azure and post it here if you wanted to check if you can see a better solution.

Reply:
I don't think that's necessary, that answers my question, thanks!